src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs

```diff
@@ -174,5 +216,74 @@ private static bool TryParseArguments(string[] args, out (Dictionary<string, str
+        /// <summary>
+        /// Logs an error to the specified <see cref="TextWriter" />.
+        /// </summary>
+        /// <param name="errorWriter">The <see cref="TextWriter" /> to write the error to.</param>
+        /// <param name="format">The formatted string of the error.</param>
+        /// <param name="args">An object array of zero or more objects to format with the error message.</param>
+        /// <returns><see langword="false" /></returns>
+        private static void LogError(TextWriter errorWriter, string format, params object[] args)
+        {
+            errorWriter.WriteLine(format, args);
+        }
+
+        /// <summary>
+        /// Attempts to deserialize global properties from the standard input stream.
+        /// </summary>
+        /// <remarks>
+        /// <param name="errorWriter">A <see cref="TextWriter" /> to write errors to if one occurs.</param>
+        /// <param name="reader">The <see cref="BinaryReader" /> to use when deserializing the arguments.</param>
+        /// <param name="globalProperties">Receives a <see cref="Dictionary{TKey, TValue}" /> representing the global properties.</param>
+        /// <returns><see langword="true" /> if the arguments were successfully deserialized, otherwise <see langword="false" />.</returns>
+        internal static bool TryDeserializeGlobalProperties(TextWriter errorWriter, BinaryReader reader, out Dictionary<string, string> globalProperties)
+        {
+            globalProperties = null;
+
+            int count = 0;
+
+            try
+            {
+                // Read the first integer from the stream which is the number of global properties
+                count = reader.ReadInt32();
+            }
+            catch (Exception e)
+            {
+                LogError(errorWriter, Strings.Error_StaticGraphRestoreArgumentsParsingFailedExceptionReadingStream, e.Message, e.ToString());
+
+                return false;
+            }
+
+            // If the integer is negative or greater than or equal to int.MaxValue, then the integer is invalid.  This should never happen unless the bytes in the stream contain completely unexpected values
+            if (count < 0 || count >= int.MaxValue)
+            {
+                // An error occurred parsing command-line arguments in static graph-based restore as the first integer read, {0}, was is greater than the allowable value. Please file an issue at https://github.com/NuGet/Home
+                LogError(errorWriter, Strings.Error_StaticGraphRestoreArgumentsParsingFailedUnexpectedIntegerValue, count);
+
+                return false;
+            }
+
+            globalProperties = new Dictionary<string, string>(capacity: count, StringComparer.OrdinalIgnoreCase);
+
+            for (int i = 0; i < count; i++)
+            {
+                try
+                {
+                    globalProperties[reader.ReadString()] = reader.ReadString();
+                }
+                catch (Exception e)
```

Comment:
Should we invoke Close() method to close the current reader and the unederlying stream either in the catch block or finally block?